main.go

```diff
@@ -0,0 +1,195 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/printers"
+
+	bootstrapapi "k8s.io/cluster-bootstrap/token/api"
+	bootstraputil "k8s.io/cluster-bootstrap/token/util"
+	kubeadmapiv1 "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta3"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/c3os-io/c3os/sdk/clusterplugin"
+	yip "github.com/mudler/yip/pkg/schema"
+	"github.com/sirupsen/logrus"
+	bootstraptokenv1 "k8s.io/kubernetes/cmd/kubeadm/app/apis/bootstraptoken/v1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	_ = kubeadmapiv1.AddToScheme(scheme)
+}
+
+var configurationPath = "/opt/kubeadm"
+
+func main() {
+	plugin := clusterplugin.ClusterPlugin{
+		Provider: clusterProvider,
+	}
+
+	if err := plugin.Run(); err != nil {
+		logrus.Fatal(err)
+	}
+}
+
+func clusterProvider(cluster clusterplugin.Cluster) yip.YipConfig {
+
+	var stages []yip.Stage
+
+	preStage := yip.Stage{
+		Systemctl: yip.Systemctl{
+			Enable: []string{"kubelet"},
+			Start:  []string{"containerd"},
+		},
+		Commands: []string{
+			"sysctl --system",
+			"modprobe overlay",
+			"modprobe br_netfilter",
+		},
+	}
+
+	if !bootstraputil.IsValidBootstrapToken(cluster.ClusterToken) {
+		logrus.Fatalf("the bootstrap token %s is not of the form %s", cluster.ClusterToken, bootstrapapi.BootstrapTokenPattern)
+		return yip.YipConfig{}
+	}
+
+	stages = append(stages, preStage)
+
+	if cluster.Role == clusterplugin.RoleInit {
+		stages = append(stages, getInitYipStages(cluster)...)
+	} else if (cluster.Role == clusterplugin.RoleControlPlane) || (cluster.Role == clusterplugin.RoleWorker) {
+		stages = append(stages, getJoinYipStages(cluster)...)
+	}
+
+	cfg := yip.YipConfig{
+		Name: "Kubeadm C3OS Cluster Provider",
+		Stages: map[string][]yip.Stage{
+			"boot.before": stages,
+		},
+	}
+
+	return cfg
+}
+
+func getInitYipStages(cluster clusterplugin.Cluster) []yip.Stage {
+	kubeadmCfg := getInitNodeConfiguration(cluster)
+	return []yip.Stage{
+		{
+			Name: "Init Kubeadm",
+			If:   "[ ! -f /opt/kubeadm.init ]",
+			Files: []yip.File{
+				{
+					Path:        filepath.Join(configurationPath, "kubeadm.yaml"),
+					Permissions: 0640,
+					Content:     kubeadmCfg,
+				},
+			},
+			Commands: []string{
+				fmt.Sprintf("kubeadm init --config %s --upload-certs", filepath.Join(configurationPath, "kubeadm.yaml")),
+			},
+		},
+	}
+}
+
+func getJoinYipStages(cluster clusterplugin.Cluster) []yip.Stage {
+	kubeadmCfg := getJoinNodeConfiguration(cluster)
+
+	joinCmd := fmt.Sprintf("kubeadm join --config %s --ignore-preflight-errors=DirAvailable--etc-kubernetes-manifests", filepath.Join(configurationPath, "kubeadm.yaml"))
+	if cluster.Role == clusterplugin.RoleControlPlane {
+		joinCmd = joinCmd + " --control-plane"
+	}
+
+	return []yip.Stage{
+		{
+			Name: "Join Kubeadm",
+			If:   "[ ! -f /opt/kubeadm.join ]",
```

Comment:
@kpiyush17 when will this file be created ? /opt/kubeadm.join

Reply:
@jzhoucliqr It seems I am not creating this file now. I will add it. This file will be created as part of the join command and the same for the init command.
This was in accordance with this comment: https://github.com/kairos-io/provider-kubeadm/pull/1/commits/03cba3a3409f2713a69c6c71a4d7ef02e493af2f#r968737034

Reply:
Creating the files as part of kubeadm init/join commands